### DISCUSSION TITLE
[403 Forbidden] submit a notice

### DISCUSSION THREAD

**Discussion by ismailatitallah** [OP-TED/eForms-SDK#255] [Q&A] [answered]
Hello,

I am trying to submit a notice with https://enotices2.preview.ted.europa.eu/ api. I created my api-key from the web-site https://developer.preview.ted.europa.eu/ted-api . And I added this api-key to my api call in the header : "Bearer xxxxxxx". But I am getting 403 Forbidden as response.

**Comment by pcolast** [comment, marked as answer]
Hello,

Before starting using the API, it is needed to log into eNotices2, so that eNotices2 account is created.
The same EU login needs to be used as used to generated the TEDAPI key.

Regards,
Stefan

**Reply by ismailatitallah** [reply]
I already done that.

But when I send it with curl command, i have 403 Forbidden.
this is my curl :
```
curl -X 'POST' \
  'https://enotices2.ted.europa.eu/esenders/api/v2/notice/notices/submit' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer xxxxxxxxxxxxxxxxxxxxx' \
  -H 'Content-Type: multipart/form-data' \
  -F 'notice=@notice-16.xml;type=text/xml' \
  -F 'esenderAuthor={
  "noticeAuthorEmail": "john@doe.com",
  "noticeAuthorLocale": "en"
}'
```
But this is the response

`<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<TITLE>ERROR: The request could not be satisfied</TITLE>
</HEAD><BODY>
<H1>403 ERROR</H1>
<H2>The request could not be satisfied.</H2>
<HR noshade size="1px">
Request blocked.
We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error. Try again later, or contact the app or website owner.
<BR clear="all">
If you provide content to customers through CloudFront, you can find steps to troubleshoot and help prevent this error by reviewing the CloudFront documentation.
<BR clear="all">
<HR noshade size="1px">
<PRE>
Generated by cloudfront (CloudFront)
Request ID: PigFgw_00oIfY03-F89bozGe5ZPidsTg_0xHbh3_DGho2jsegTBcIQ==
</PRE>
<ADDRESS>
</ADDRESS>
</BODY></HTML>`

**Reply by pcolast** [reply]
You need to have a TEDAPI key for Preview and use it only in Preview, for PROD you need to create TEDAPI key in https://developer.ted.europa.eu/ted-api

Your last CURL was from PROD, how about Preview, is the result the same ?

**Reply by ismailatitallah** [reply]
Thanks, it's much better. 
Now I am having 
`{"code":"notice.SDK_NOT_FOUND","message":"No compatible SDK version found for featureVersion=eforms-sdk-1.4","showDetails":false}`
what sdk version do you recommand me to use, in order to have response 200?

**Reply by pcolast** [reply]
You are welcome!
Please always consult the Preview page for the latest compatible SDK version: [eForms Preview environment :: TED Developer Documentation (europa.eu)](https://docs.ted.europa.eu/home/eforms/preview/index.html#_enotices2)

**Reply by ismailatitallah** [reply]
The SDK available from your web site is 1.3.2. In this version, in fields.json, there is an attribute sdkVersion with the value "eforms-sdk-1.3.2". When I use this value to build my xml, i had an error when sybmitting `{"code":"notice.SDK_NOT_FOUND","message":"No compatible SDK version found for featureVersion=eforms-sdk-1.3.2","showDetails":false}` What am I doing wrong?
 
![image](https://user-images.githubusercontent.com/78085225/206396048-030a3daa-ab3a-4278-8c67-3e740f643056.png)

**Reply by bertrand-lorentz** [reply]
For the value to indicate in the `cbc:CustomizationID` element, it should always have the format "eforms-sdk-_major.minor_". So in your case: "eforms-sdk-1.3"
See this page for more details:
https://docs.ted.europa.eu/eforms/latest/versioning.html#_significance_of_the_sdk_version_in_notice_handling_and_validation

**Comment by gediminasre** [comment]
Hi, I have an issue with the same (i think) endpoint. https://enotices2.preview.ted.europa.eu/esenders/webjars/swagger-ui/index.html#/Notice/submitNotice
It is working fine when uploading via Swagger or curl, but we are struggling to send same notice via code. We keep getting 400 Bad request, but not sure what we are doing bad.
Can you share the part of code this endpoint is expecting? So we can adjust our code.
Our code structure is basically this, plus headers accept: application/json, Authorization: Bearer xxXXxx, Content-Type: multipart/form-data:
```
public class RequestBody {
  @FormParam("notice")
  @PartType(MediaType.TEXT_XML)
  private byte[] notice;

  @FormParam("esenderAuthor")
  @PartType(MediaType.APPLICATION_JSON)
  private Author esenderAuthor;
}

  static class Author {
  private String noticeAuthorEmail = "email@email.com";
  private String noticeAuthorLocale = "en";
  }
```

**Reply by pcolast** [reply]
Apart from the HTTP code, could you please provide what is returned in the response body ?

**Reply by gediminasre** [reply]
No response body is returned. Attaching a screenshot:
![image](https://user-images.githubusercontent.com/118281897/207014379-aac5ab1b-b29e-4558-ade3-0fd45e52129f.png)
P.S. We definitely have a connection with TED, since we can successfully get correct responses from /notices/search-esentool endpoint.

**Comment by gediminasre** [comment]
Solved. For anyone else facing this issue, make sure your requests' header "Content-Disposition" contains filename property.
i.e. Content-Disposition: form-data; name="notice"; **filename**="anyfilename.xml"

**Reply by pcolast** [reply]
Many thanks for your feedback.
